src/data/equiv/local_equiv.lean

```diff
@@ -0,0 +1,518 @@
+  right_inv  := λx hx, e.right_inv x }
+
+namespace local_equiv
+
+variables (e : local_equiv α β) (e' : local_equiv β γ)
+
+/-- Associating to a local_equiv an equiv between the source and the target -/
+protected def to_equiv : equiv (e.source) (e.target) :=
+{ to_fun    := λ⟨x, hx⟩, ⟨e.to_fun x, e.map_source hx⟩,
+  inv_fun   := λ⟨y, hy⟩, ⟨e.inv_fun y, e.map_target hy⟩,
+  left_inv  := λ⟨x, hx⟩, subtype.eq $ e.left_inv hx,
+  right_inv := λ⟨y, hy⟩, subtype.eq $ e.right_inv hy }
+
+/-- The inverse of a local equiv -/
+protected def symm : local_equiv β α :=
+{ to_fun     := e.inv_fun,
+  inv_fun    := e.to_fun,
+  source     := e.target,
+  target     := e.source,
+  map_source := e.map_target,
+  map_target := e.map_source,
+  left_inv   := e.right_inv,
+  right_inv  := e.left_inv }
+
+@[simp] lemma symm_to_fun : e.symm.to_fun = e.inv_fun := rfl
+@[simp] lemma symm_inv_fun : e.symm.inv_fun = e.to_fun := rfl
+@[simp] lemma symm_source : e.symm.source = e.target := rfl
+@[simp] lemma symm_target : e.symm.target = e.source := rfl
+@[simp] lemma symm_symm : e.symm.symm = e := by { cases e, refl }
+
+/-- A local equiv induces a bijection between its source and target -/
+lemma bij_on_source : bij_on e.to_fun e.source e.target :=
+bij_on_of_inv_on e.map_source e.map_target ⟨e.left_inv, e.right_inv⟩
+
+lemma image_eq_target_inter_inv_preimage {s : set α} (h : s ⊆ e.source) :
+  e.to_fun '' s = e.target ∩ e.inv_fun ⁻¹' s :=
+begin
+  refine subset.antisymm (λx hx, _) (λx hx, _),
+  { rcases (mem_image _ _ _).1 hx with ⟨y, ys, hy⟩,
+    rw ← hy,
+    split,
+    { apply e.map_source,
+      exact h ys },
+    { rwa [mem_preimage, e.left_inv (h ys)] } },
+  { rw ← e.right_inv hx.1,
+    exact mem_image_of_mem _ hx.2 }
+end
+
+lemma inv_image_eq_source_inter_preimage {s : set β} (h : s ⊆ e.target) :
+  e.inv_fun '' s = e.source ∩ e.to_fun ⁻¹' s :=
+e.symm.image_eq_target_inter_inv_preimage h
+
+lemma source_inter_preimage_inv_preimage (s : set α) :
+  e.source ∩ e.to_fun ⁻¹' (e.inv_fun ⁻¹' s) = e.source ∩ s :=
+begin
+  ext, split,
+  { rintros ⟨hx, xs⟩,
+    simp only [mem_preimage, hx, e.left_inv, mem_preimage] at xs,
+    exact ⟨hx, xs⟩ },
+  { rintros ⟨hx, xs⟩,
+    simp [hx, xs] }
+end
+
+lemma target_inter_inv_preimage_preimage (s : set β) :
+  e.target ∩ e.inv_fun ⁻¹' (e.to_fun ⁻¹' s) = e.target ∩ s :=
+e.symm.source_inter_preimage_inv_preimage _
+
+lemma image_source_eq_target : e.to_fun '' e.source = e.target :=
+image_eq_of_bij_on e.bij_on_source
+
+lemma source_subset_preimage_target : e.source ⊆ e.to_fun ⁻¹' e.target :=
+λx hx, e.map_source hx
+
+lemma inv_image_target_eq_source : e.inv_fun '' e.target = e.source :=
+image_eq_of_bij_on e.symm.bij_on_source
+
+lemma target_subset_preimage_source : e.target ⊆ e.inv_fun ⁻¹' e.source :=
+λx hx, e.map_target hx
+
+/-- Two local equivs that have the same source, same to_fun and same inv_fun, coincide. -/
+protected lemma eq (e' : local_equiv α β) (h : ∀x, e.to_fun x = e'.to_fun x)
```

Comment:
Do you want to label this with `@[extensionality]`, so `ext` can automatically apply it?